src/cryptography/hazmat/backends/openssl/ec.py

```diff
@@ -182,6 +182,11 @@ def exchange(self, algorithm, peer_public_key):
                 _Reasons.UNSUPPORTED_EXCHANGE_ALGORITHM
             )
 
+        if type(peer_public_key.curve) is not type(self.curve):
```

Comment:
Yeah, that would probably look cleaner. When writing an __eq__ for classes that have other members to compare, part of it is usually involves an isinstance() check of this sort before comparing the members. In this case, if these classes have no other state, the isinstance() by itself could  be enough.

That said, is there any reason these are classes instead of instances that could just be compared directly with "is"? Is it just for consistency with other parts of the API where arguments might be passed when constructing such an object?

One possibility here would be to make these singletons, so you still construct them from classes, but the constructor always returns the same instance.